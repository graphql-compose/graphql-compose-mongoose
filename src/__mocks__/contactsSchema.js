/* @flow */

import { Schema } from './mongooseCommon';

const ContactsSchema = new Schema({
  phones: [String],
  email: String,
  skype: String,
  locationId: Schema.Types.ObjectId,
});

export default ContactsSchema;
