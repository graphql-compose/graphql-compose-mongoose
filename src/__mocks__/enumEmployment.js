/* @flow */

const enumEmployment = {
  full: { description: 'Full time' },
  partial: { description: 'Partial time' },
  remote: { description: 'Remote work' },
};

export default enumEmployment;
