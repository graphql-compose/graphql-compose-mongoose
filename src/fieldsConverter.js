/* @flow */
/* eslint-disable no-use-before-define */

import mongoose from 'mongoose';
import objectPath from 'object-path';
import {
  TypeComposer,
  GraphQLDate,
  GraphQLBuffer,
  GraphQLGeneric,
  GraphQLJSON,
} from 'graphql-compose';

import {
  GraphQLString,
  GraphQLFloat,
  GraphQLBoolean,
  GraphQLList,
  GraphQLEnumType,
  GraphQLObjectType,
} from 'graphql';
import GraphQLMongoID from './types/mongoid';
import typeStorage from './typeStorage';

import type {
  MongooseModelT,
  MongooseModelSchemaT,
  MongoosePseudoModelT,
  MongooseFieldT,
  MongooseFieldMapT,
  ComplexTypesT,
  GraphQLOutputType,
} from './definition';

export const ComplexTypes = {
  ARRAY: 'ARRAY',
  EMBEDDED: 'EMBEDDED',
  DOCUMENT_ARRAY: 'DOCUMENT_ARRAY',
  ENUM: 'ENUM',
  REFERENCE: 'REFERENCE',
  SCALAR: 'SCALAR',
  MIXED: 'MIXED',
};

function _getFieldName(field: MongooseFieldT): string {
  return field.path;
}

function _getFieldType(field: MongooseFieldT): string {
  return field.instance;
}

function _getFieldDescription(field: MongooseFieldT): ?string {
  if (field.options && field.options.description) {
    return field.options.description;
  }

  return undefined;
}

function _getFieldEnums(field: MongooseFieldT): ?string[] {
  if (field.enumValues && field.enumValues.length > 0) {
    return field.enumValues;
  }

  return undefined;
}

function capitalize(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function dotPathsToEmbedded(fields: MongooseFieldMapT): MongooseFieldMapT {
  // convert only one dot-level on this step to EmbeddedModel
  // further when converting EmbeddedModel to GQL, it internally
  // call this method to extract deep fields with dots

  const result: MongooseFieldMapT = {};

  Object.keys(fields).forEach((fieldName) => {
    const dotIdx = fieldName.indexOf('.');
    if (dotIdx === -1) {
      result[fieldName] = fields[fieldName];
    } else {
      // create pseudo sub-model
      const name = fieldName.substr(0, dotIdx);
      if (!result[name]) {
        const embeddedField: MongooseFieldT = {
          instance: 'Embedded',
          path: name,
          schema: {
            paths: {},
          },
        };
        result[name] = embeddedField;
      }
      const subName = fieldName.substr(dotIdx + 1);
      result[name].schema.paths[subName] = Object.assign({}, fields[fieldName], { path: subName });
    }
  });

  return result;
}

export function getFieldsFromModel(
  model: MongooseModelT | MongoosePseudoModelT
): MongooseFieldMapT {
  if (!model || !model.schema || !model.schema.paths) {
    throw new Error('You provide incorrect mongoose model to `getFieldsFromModel()`. '
    + 'Correct model should contain `schema.paths` properties.');
  }

  const fields = {};
  const paths = dotPathsToEmbedded(model.schema.paths);

  Object
    .keys(paths)
    .filter(path => !path.startsWith('__')) // skip hidden fields
    .forEach((path) => {
      fields[path] = paths[path];
    });

  return fields;
}

export function convertModelToGraphQL(
  model: MongooseModelT | MongoosePseudoModelT,
  typeName: string
): TypeComposer {
  // if model already has generated TypeComposer early, then return it
  if (model.schema && model.schema._gqcTypeComposer instanceof TypeComposer) {
    return model.schema._gqcTypeComposer;
  }

  if (!typeName) {
    throw new Error('You provide empty name for type. '
    + '`name` argument should be non-empty string.');
  }

  const typeComposer = new TypeComposer(
    typeStorage.getOrSet(
      typeName,
      new GraphQLObjectType({
        name: typeName,
        interfaces: [],
        description: undefined,
        fields: {},
      })
    )
  );
  model.schema._gqcTypeComposer = typeComposer; // eslint-disable-line

  const mongooseFields = getFieldsFromModel(model, typeName);
  const graphqlFields = {};

  Object.keys(mongooseFields).forEach((fieldName) => {
    const mongooseField:MongooseFieldT = mongooseFields[fieldName];
    graphqlFields[fieldName] = {
      type: convertFieldToGraphQL(mongooseField, typeName),
      description: _getFieldDescription(mongooseField),
    };

    if (deriveComplexType(mongooseField) === ComplexTypes.EMBEDDED) {
      // https://github.com/nodkz/graphql-compose-mongoose/issues/7
      graphqlFields[fieldName].resolve = (source) => {
        if (source) {
          if (source.toObject) {
            const obj = source.toObject();
            return obj[fieldName];
          }
          return source[fieldName];
        }
        return null;
      };
    }
  });

  typeComposer.addFields(graphqlFields);
  return typeComposer;
}

export function convertSchemaToGraphQL(
  schema: MongooseModelSchemaT,
  typeName: string
) {
  if (!typeName) {
    throw new Error('You provide empty name for type. '
    + '`name` argument should be non-empty string.');
  }

  if (schema._gqcTypeComposer) {
    return schema._gqcTypeComposer;
  }

  const tc = convertModelToGraphQL({ schema }, typeName);
  // also generate InputType
  tc.getInputTypeComposer();

  schema._gqcTypeComposer = tc; // eslint-disable-line
  return tc;
}

export function convertFieldToGraphQL(
  field: MongooseFieldT,
  prefix: string = ''
): GraphQLOutputType {
  const complexType = deriveComplexType(field);
  switch (complexType) {
    case ComplexTypes.SCALAR: return scalarToGraphQL(field);
    case ComplexTypes.ARRAY: return arrayToGraphQL(field, prefix);
    case ComplexTypes.EMBEDDED: return embeddedToGraphQL(field, prefix);
    case ComplexTypes.ENUM: return enumToGraphQL(field, prefix);
    case ComplexTypes.REFERENCE: return referenceToGraphQL(field, prefix);
    case ComplexTypes.DOCUMENT_ARRAY: return documentArrayToGraphQL(field, prefix);
    case ComplexTypes.MIXED: return mixedToGraphQL(field);
    default: return scalarToGraphQL(field);
  }
}

export function deriveComplexType(field: MongooseFieldT): ComplexTypesT {
  if (!field || !field.path || !field.instance) {
    throw new Error('You provide incorrect mongoose field to `deriveComplexType()`. '
    + 'Correct field should contain `path` and `instance` properties.');
  }

  const fieldType = _getFieldType(field);

  if (field instanceof mongoose.Schema.Types.DocumentArray) {
    return ComplexTypes.DOCUMENT_ARRAY;
  } else if (field instanceof mongoose.Schema.Types.Embedded
    || fieldType === 'Embedded'
    ) {
    return ComplexTypes.EMBEDDED;
  } else if (field instanceof mongoose.Schema.Types.Array
    || objectPath.has(field, 'caster.instance')) {
    return ComplexTypes.ARRAY;
  } else if (field instanceof mongoose.Schema.Types.Mixed) {
    return ComplexTypes.MIXED;
  } else if (fieldType === 'ObjectID') {
    return ComplexTypes.REFERENCE;
  }

  const enums = _getFieldEnums(field);
  if (enums) {
    return ComplexTypes.ENUM;
  }

  return ComplexTypes.SCALAR;
}

export function scalarToGraphQL(field: MongooseFieldT): GraphQLOutputType {
  const typeName = _getFieldType(field);

  switch (typeName) {
    case 'String': return GraphQLString;
    case 'Number': return GraphQLFloat;
    case 'Date': return GraphQLDate;
    case 'Buffer': return GraphQLBuffer;
    case 'Boolean': return GraphQLBoolean;
    case 'ObjectID': return GraphQLMongoID;
    default: return GraphQLGeneric;
  }
}

export function arrayToGraphQL(field: MongooseFieldT, prefix: string = ''): GraphQLOutputType {
  if (!field || !field.caster) {
    throw new Error('You provide incorrect mongoose field to `arrayToGraphQL()`. '
    + 'Correct field should contain `caster` property.');
  }

  const unwrappedField = Object.assign({}, field.caster);
  objectPath.set(
    unwrappedField,
    'options.ref',
    objectPath.get(field, 'options.ref', undefined)
  );

  const outputType = convertFieldToGraphQL(unwrappedField, prefix);
  return new GraphQLList(outputType);
}

export function embeddedToGraphQL(
  field: MongooseFieldT,
  prefix: string = ''
): GraphQLOutputType {
  const fieldName = _getFieldName(field);
  const fieldType = _getFieldType(field);

  if (fieldType !== 'Embedded') {
    throw new Error('You provide incorrect field to `embeddedToGraphQL()`. '
    + 'This field should has `Embedded` type. ');
  }

  const typeName = `${prefix}${capitalize(fieldName)}`;
  const typeComposer = convertSchemaToGraphQL(field.schema, typeName);

  return typeComposer.getType();
}

export function enumToGraphQL(
  field: MongooseFieldT,
  prefix: string = ''
): GraphQLOutputType {
  const valueList = _getFieldEnums(field);
  if (!valueList) {
    throw new Error('You provide incorrect mongoose field to `enumToGraphQL()`. '
    + 'Correct field should contain `enumValues` property');
  }

  const graphQLEnumValues = valueList.reduce((result, val) => {
    result[val] = { value: val }; // eslint-disable-line no-param-reassign
    return result;
  }, {});

  const typeName = `Enum${prefix}${capitalize(_getFieldName(field))}`;
  return typeStorage.getOrSet(
    typeName,
    new GraphQLEnumType({
      name: typeName,
      description: _getFieldDescription(field),
      values: graphQLEnumValues,
    })
  );
}


export function documentArrayToGraphQL(
  field: MongooseFieldT,
  prefix: string = ''
): GraphQLOutputType {
  if (!(field instanceof mongoose.Schema.Types.DocumentArray)) {
    throw new Error('You provide incorrect mongoose field to `documentArrayToGraphQL()`. '
    + 'Correct field should be instance of `mongoose.Schema.Types.DocumentArray`');
  }

  const typeName = `${prefix}${capitalize(_getFieldName(field))}`;

  const typeComposer = convertModelToGraphQL(field, typeName);

  return new GraphQLList(typeComposer.getType());
}


export function referenceToGraphQL(
  field: MongooseFieldT,
  prefix: string = ''
): GraphQLOutputType {
  const fieldType = _getFieldType(field);
  if (fieldType !== 'ObjectID') {
    throw new Error('You provide incorrect mongoose field to `referenceToGraphQL()`. '
    + 'Correct field should has mongoose-type `ObjectID`');
  }

  // const refModelName = objectPath.get(field, 'options.ref');
  // if (refModelName) {
  //   return GQLReference;
  //   // throw new Error('Mongoose REFERENCE to graphQL TYPE not implemented yet. '
  //   //                + `Field ${_getFieldName(field)}`);
  //   // Storage.UnresolvedRefs.setSubKey(parentTypeName, fieldName, { refModelName });
  //   // return GraphQLReference;
  // }

  // this is mongo id field
  return scalarToGraphQL(field, prefix);
}

export function mixedToGraphQL(
  field: MongooseFieldT
): GraphQLOutputType {
  if (!(field instanceof mongoose.Schema.Types.Mixed)) {
    throw new Error('You provide incorrect mongoose field to `mixedToGraphQL()`. '
    + 'Correct field should be instance of `mongoose.Schema.Types.Mixed`');
  }

  return GraphQLJSON;
}
