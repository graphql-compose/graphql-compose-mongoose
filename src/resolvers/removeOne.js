/* @flow */
/* eslint-disable no-param-reassign */

import { Resolver, TypeComposer } from 'graphql-compose';
import GraphQLMongoID from '../types/mongoid';
import typeStorage from '../typeStorage';
import { filterHelperArgs } from './helpers/filter';
import { sortHelperArgs } from './helpers/sort';
import findOne from './findOne';
import type { MongooseModelT, ExtendedResolveParams, GenResolverOpts } from '../definition';

export default function removeOne(
  model: MongooseModelT,
  typeComposer: TypeComposer,
  opts?: GenResolverOpts
): Resolver<*, *> {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver removeOne() should be instance of Mongoose Model.');
  }

  if (!(typeComposer instanceof TypeComposer)) {
    throw new Error('Second arg for Resolver removeOne() should be instance of TypeComposer.');
  }

  const findOneResolver = findOne(model, typeComposer, opts);

  const outputTypeName = `RemoveOne${typeComposer.getTypeName()}Payload`;
  const outputType = typeStorage.getOrSet(
    outputTypeName,
    TypeComposer.create({
      name: outputTypeName,
      fields: {
        recordId: {
          type: GraphQLMongoID,
          description: 'Removed document ID',
        },
        record: {
          type: typeComposer.getType(),
          description: 'Removed document',
        },
      },
    })
  );

  const resolver = new Resolver({
    name: 'removeOne',
    kind: 'mutation',
    description:
      'Remove one document: ' +
      '1) Remove with hooks via findOneAndRemove. ' +
      '2) Return removed document.',
    type: outputType,
    args: {
      ...filterHelperArgs(typeComposer, model, {
        filterTypeName: `FilterRemoveOne${typeComposer.getTypeName()}Input`,
        model,
        ...(opts && opts.filter),
      }),
      ...sortHelperArgs(model, {
        sortTypeName: `SortRemoveOne${typeComposer.getTypeName()}Input`,
        ...(opts && opts.sort),
      }),
    },
    resolve: (resolveParams: ExtendedResolveParams) => {
      const filterData = (resolveParams.args && resolveParams.args.filter) || {};

      if (!(typeof filterData === 'object') || Object.keys(filterData).length === 0) {
        return Promise.reject(
          new Error(
            `${typeComposer.getTypeName()}.removeOne resolver requires ` +
              'at least one value in args.filter'
          )
        );
      }

      // We should get all data for document, cause Mongoose model may have hooks/middlewares
      // which required some fields which not in graphql projection
      // So empty projection returns all fields.
      resolveParams.projection = {};

      return (
        findOneResolver
          .resolve(resolveParams)
          .then(doc => {
            if (resolveParams.beforeRecordMutate) {
              return resolveParams.beforeRecordMutate(doc, resolveParams);
            }
            return doc;
          })
          // remove record from DB
          .then(doc => {
            if (!doc) {
              return Promise.reject(new Error('Document not found'));
            }
            return doc.remove();
          })
          // prepare output payload
          .then(record => {
            if (record) {
              return {
                record,
                recordId: typeComposer.getRecordIdFn()(record),
              };
            }

            return null;
          })
      );
    },
  });

  return resolver;
}
