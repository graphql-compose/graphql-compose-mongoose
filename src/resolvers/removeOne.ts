import type { Resolver, ObjectTypeComposer } from 'graphql-compose';
import type { Model, Document } from 'mongoose';
import {
  filterHelperArgs,
  sortHelperArgs,
  SortHelperArgsOpts,
  FilterHelperArgsOpts,
} from './helpers';
import { findOne } from './findOne';
import type { ExtendedResolveParams } from './index';
import { addErrorCatcherField } from './helpers/errorCatcher';
import { payloadRecordId, PayloadRecordIdHelperOpts } from './helpers/payloadRecordId';

export interface RemoveOneResolverOpts {
  /** If you want to generate different resolvers you may avoid Type name collision by adding a suffix to type names */
  suffix?: string;
  /** Customize input-type for `filter` argument. If `false` then arg will be removed. */
  filter?: FilterHelperArgsOpts | false;
  sort?: SortHelperArgsOpts | false;
  /** Customize payload.recordId field. If false, then this field will be removed. */
  recordId?: PayloadRecordIdHelperOpts | false;
}

type TArgs = {
  filter?: any;
  sort?: Record<string, any>;
};

export function removeOne<TSource = any, TContext = any, TDoc extends Document = any>(
  model: Model<TDoc>,
  tc: ObjectTypeComposer<TDoc, TContext>,
  opts?: RemoveOneResolverOpts
): Resolver<TSource, TContext, TArgs, TDoc> {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver removeOne() should be instance of Mongoose Model.');
  }

  if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
    throw new Error(
      'Second arg for Resolver removeOne() should be instance of ObjectTypeComposer.'
    );
  }

  const findOneResolver = findOne(model, tc, opts);

  const outputTypeName = `RemoveOne${tc.getTypeName()}${opts?.suffix || ''}Payload`;
  const outputType = tc.schemaComposer.getOrCreateOTC(outputTypeName, (t) => {
    t.setFields({
      ...payloadRecordId(tc, opts?.recordId),
      record: {
        type: tc,
        description: 'Removed document',
      },
    });
  });

  const resolver = tc.schemaComposer.createResolver<TSource, TArgs>({
    name: 'removeOne',
    kind: 'mutation',
    description:
      'Remove one document: ' +
      '1) Remove with hooks via findOneAndRemove. ' +
      '2) Return removed document.',
    type: outputType,
    args: {
      ...filterHelperArgs(tc, model, {
        prefix: 'FilterRemoveOne',
        suffix: `${opts?.suffix || ''}Input`,
        ...opts?.filter,
      }),
      ...sortHelperArgs(tc, model, {
        sortTypeName: `SortRemoveOne${tc.getTypeName()}${opts?.suffix || ''}Input`,
        ...opts?.sort,
      }),
    },
    resolve: (async (resolveParams: ExtendedResolveParams<TDoc>) => {
      const filterData = resolveParams?.args?.filter;

      if (!(typeof filterData === 'object') || Object.keys(filterData).length === 0) {
        return Promise.reject(
          new Error(
            `${tc.getTypeName()}.removeOne resolver requires at least one value in args.filter`
          )
        );
      }

      // We should get all data for document, cause Mongoose model may have hooks/middlewares
      // which required some fields which not in graphql projection
      // So empty projection returns all fields.
      resolveParams.projection = {};

      let doc: TDoc = await findOneResolver.resolve(resolveParams);

      if (resolveParams.beforeRecordMutate) {
        doc = await resolveParams.beforeRecordMutate(doc, resolveParams);
      }

      if (doc) {
        await doc.remove();

        return {
          record: doc,
        };
      }

      return null;
    }) as any,
  });

  // Add `error` field to payload which can catch resolver Error
  // and return it in mutation payload
  addErrorCatcherField(resolver);

  return resolver;
}
