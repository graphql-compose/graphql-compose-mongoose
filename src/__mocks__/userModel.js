/* @flow */

import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'UserModel',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'UserModel',
    },

    name: {
      type: String,
      description: 'Person name',
    },

    age: {
      type: Number,
      description: 'Full years',
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    // createdAt, created via option `timastamp: true` (see bottom)
    // updatedAt, created via option `timastamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

UserSchema.virtual('nameVirtual').get(function () { // eslint-disable-line
  return `VirtualFieldValue${this._id}`;
});

const UserModel = mongoose.model('User', UserSchema);

export { UserSchema, UserModel };
