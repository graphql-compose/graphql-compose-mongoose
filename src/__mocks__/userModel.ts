import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';
import { Document } from 'mongoose';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'User',
    },

    n: {
      type: String,
      required: true,
      description: 'Person name',
      alias: 'name',
    },

    age: {
      type: Number,
      description: 'Full years',
      required() {
        // in graphql this field should be Nullable
        return (this as any).name === 'Something special';
      },
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    periods: [{ from: Number, to: Number }],

    someDeep: {
      periods: [{ from: Number, to: Number }],
    },

    salary: {
      type: Schema.Types.Decimal128,
    },

    mapField: {
      type: Map,
      of: String,
    },

    mapFieldDeep: {
      subField: {
        type: Map,
        of: String,
      },
    },

    billingAddress: {
      street: { type: String, index: true },
      state: { type: String, index: true, enum: ['FL', 'MA', 'NY'] },
      country: { type: String, index: true },
    },

    // for error payloads tests
    valid: {
      type: String,
      required: false,
      validate: [
        () => {
          return false;
        },
        'this is a validate message',
      ],
    },

    // createdAt, created via option `timestamp: true` (see bottom)
    // updatedAt, created via option `timestamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

// eslint-disable-next-line
UserSchema.virtual('nameVirtual').get(function (this: any) {
  return `VirtualFieldValue${this._id}`;
});

export interface IUser extends Document {
  _id: any;
  name?: string;
  age?: number;
  gender?: string;
  someDynamic?: any;
  skills?: string[];
  relocation?: boolean;
  contacts: {
    email: string;
    skype?: string;
  };
}

const UserModel = mongoose.model<IUser>('User', UserSchema);

export { UserSchema, UserModel };
