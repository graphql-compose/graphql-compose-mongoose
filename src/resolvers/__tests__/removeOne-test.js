/* @flow */
/* eslint-disable no-param-reassign */

import { Resolver, schemaComposer } from 'graphql-compose';
import { UserModel } from '../../__mocks__/userModel';
import removeOne from '../removeOne';
import GraphQLMongoID from '../../types/mongoid';
import { convertModelToGraphQL } from '../../fieldsConverter';

beforeAll(() => UserModel.base.connect());
afterAll(() => UserModel.base.disconnect());

describe('removeOne() ->', () => {
  let UserTC;

  beforeEach(() => {
    schemaComposer.clear();
    UserTC = convertModelToGraphQL(UserModel, 'User', schemaComposer);
    UserTC.setRecordIdFn(source => (source ? `${source._id}` : ''));
  });

  let user1;
  let user2;
  let user3;

  beforeEach(async () => {
    await UserModel.deleteMany({});

    user1 = new UserModel({
      name: 'userName1',
      gender: 'male',
      relocation: true,
      age: 28,
    });

    user2 = new UserModel({
      name: 'userName2',
      gender: 'female',
      relocation: true,
      age: 29,
    });

    user3 = new UserModel({
      name: 'userName3',
      gender: 'female',
      relocation: true,
      age: 30,
    });

    await Promise.all([user1.save(), user2.save(), user3.save()]);
  });

  it('should return Resolver object', () => {
    const resolver = removeOne(UserModel, UserTC);
    expect(resolver).toBeInstanceOf(Resolver);
  });

  describe('Resolver.args', () => {
    it('should have `filter` arg', () => {
      const resolver = removeOne(UserModel, UserTC);
      expect(resolver.hasArg('filter')).toBe(true);
    });

    it(
      'should not have `skip` arg due mongoose error: ' +
        'skip cannot be used with findOneAndRemove',
      () => {
        const resolver = removeOne(UserModel, UserTC);
        expect(resolver.hasArg('skip')).toBe(false);
      }
    );

    it('should have `sort` arg', () => {
      const resolver = removeOne(UserModel, UserTC);
      expect(resolver.hasArg('sort')).toBe(true);
    });
  });

  describe('Resolver.resolve():Promise', () => {
    it('should be promise', () => {
      const result = removeOne(UserModel, UserTC).resolve({});
      expect(result).toBeInstanceOf(Promise);
      result.catch(() => 'catch error if appears, hide it from mocha');
    });

    it('should return payload.recordId if record existed in db', async () => {
      const result = await removeOne(UserModel, UserTC).resolve({
        args: { filter: { _id: user1.id } },
      });
      expect(result.recordId).toBe(user1.id);
    });

    it('should remove document in database', async () => {
      const checkedName = 'nameForMongoDB';
      await removeOne(UserModel, UserTC).resolve({
        args: {
          filter: { _id: user1.id },
          input: { name: checkedName },
        },
      });

      await expect(UserModel.findOne({ _id: user1._id })).resolves.toBeNull();
    });

    it('should return payload.record', async () => {
      const result = await removeOne(UserModel, UserTC).resolve({
        args: { filter: { _id: user1.id } },
      });
      expect(result.record.id).toBe(user1.id);
    });

    it('should sort records', async () => {
      const result1 = await removeOne(UserModel, UserTC).resolve({
        args: {
          filter: { relocation: true },
          sort: { age: 1 },
        },
      });
      expect(result1.record.age).toBe(user1.age);

      const result2 = await removeOne(UserModel, UserTC).resolve({
        args: {
          filter: { relocation: true },
          sort: { age: -1 },
        },
      });
      expect(result2.record.age).toBe(user3.age);
    });

    it('should pass empty projection to findOne and got full document data', async () => {
      const result = await removeOne(UserModel, UserTC).resolve({
        args: {
          filter: { _id: user1.id },
        },
        projection: {
          record: {
            name: true,
          },
        },
      });
      expect(result.record.id).toBe(user1.id);
      expect(result.record.name).toBe(user1.name);
      expect(result.record.gender).toBe(user1.gender);
    });

    it('should return mongoose document', async () => {
      const result = await removeOne(UserModel, UserTC).resolve({
        args: { filter: { _id: user1.id } },
      });
      expect(result.record).toBeInstanceOf(UserModel);
    });

    it('should rejected with Error if args.filter is empty', async () => {
      const result = removeOne(UserModel, UserTC).resolve({ args: {} });
      await expect(result).rejects.toMatchSnapshot();
    });

    it('should call `beforeRecordMutate` method with founded `record` and `resolveParams` as args', async () => {
      let beforeMutationId;
      const result = await removeOne(UserModel, UserTC).resolve({
        args: { filter: { _id: user1.id } },
        context: { ip: '1.1.1.1' },
        beforeRecordMutate: (record, rp) => {
          beforeMutationId = record.id;
          record.someDynamic = rp.context.ip;
          return record;
        },
      });
      expect(result.record).toBeInstanceOf(UserModel);
      expect(result.record.someDynamic).toBe('1.1.1.1');
      expect(beforeMutationId).toBe(user1.id);

      const empty = await UserModel.collection.findOne({ _id: user1._id });
      expect(empty).toBe(null);
    });

    it('`beforeRecordMutate` may reject operation', async () => {
      const result = removeOne(UserModel, UserTC).resolve({
        args: { filter: { _id: user1.id } },
        context: { readOnly: true },
        beforeRecordMutate: (record, rp) => {
          if (rp.context.readOnly) {
            return Promise.reject(new Error('Denied due context ReadOnly'));
          }
          return record;
        },
      });
      await expect(result).rejects.toMatchSnapshot();
      const exist = await UserModel.collection.findOne({ _id: user1._id });
      expect(exist.name).toBe(user1.name);
    });
  });

  describe('Resolver.getType()', () => {
    it('should have correct output type name', () => {
      const outputType: any = removeOne(UserModel, UserTC).getType();
      expect(outputType.name).toBe(`RemoveOne${UserTC.getTypeName()}Payload`);
    });

    it('should have recordId field', () => {
      const outputType: any = removeOne(UserModel, UserTC).getType();
      const recordIdField = schemaComposer.createObjectTC(outputType).getFieldConfig('recordId');
      expect(recordIdField.type).toBe(GraphQLMongoID);
    });

    it('should have record field', () => {
      const outputType: any = removeOne(UserModel, UserTC).getType();
      const recordField = schemaComposer.createObjectTC(outputType).getFieldConfig('record');
      expect(recordField.type).toBe(UserTC.getType());
    });

    it('should reuse existed outputType', () => {
      const outputTypeName = `RemoveOne${UserTC.getTypeName()}Payload`;
      const existedType = schemaComposer.createObjectTC(outputTypeName);
      schemaComposer.set(outputTypeName, existedType);
      const outputType = removeOne(UserModel, UserTC).getType();
      expect(outputType).toBe(existedType.getType());
    });
  });
});
