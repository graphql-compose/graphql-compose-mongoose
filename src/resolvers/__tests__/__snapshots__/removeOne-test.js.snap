// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeOne() -> Resolver.resolve():Promise \`beforeRecordMutate\` may reject operation 1`] = `[Error: Denied due context ReadOnly]`;

exports[`removeOne() -> Resolver.resolve():Promise should rejected with Error if args.filter is empty 1`] = `[Error: User.removeOne resolver requires at least one value in args.filter]`;
